atomsci/modac/__init__.py

```diff
@@ -136,3 +165,46 @@ def download_all_files_in_collection(
                     if sub_collection_path:
                         self.download_all_files_in_collection(sub_collection_path, True)
 
+    @ensure_authenticated
+    @log_action
+    def upload_file(
+        self,
+        collection_path: str,
+        data_file_path: str,
+        attributes_file: Optional[Dict] = None,
+    ) -> bool:
+        """
+        Upload a file to the MoDaC system from local filesystem.
+
+        :param collection_path: Path where the data file should be registered in MoDaC.
+        :param data_file_path: Path to the data file on the local filesystem.
+        :param attributes_file: JSON-like dictionary containing metadata about the data file.
+        """
+        url = "/".join((self.BASE_URL, "v2", "dataObject", collection_path))
+
+        if attributes_file is None:
+            attributes_file = {}
+
+        files = {
+            "dataObjectRegistration": (
+                "attributesFile.json",
+                json.dumps(attributes_file),
+                "application/json",
+            ),
+            "dataObject": (
+                os.path.basename(data_file_path),
+                open(data_file_path, "rb"),
```

Comment:
I think your code works for 99.99% of the time, but let's say `_logger.warning` encountered a runtime error. This file would remain open. You were correct in using `finally` to catch for most error cases. However, I think it's just best practice to use the `with open(...) as f` syntax, and then you don't have to worry or think about closing:


```python
with open(data_file_path, "rb") as f:
    ...
```